Delay in updates to status index

Sebastian observed that his WARC files contained between 10-15% duplicates when crawling with ES. The most likely explanation is that there is a delay between the moment where an update is committed to the ES index and reflected in the search and when the corresponding tuple is acked (which happens a lot sooner). This issue is more likely on small crawls where the diversity of URLs is low.

One way around it is to set `es.status.min.delay.queries` to a larger value than the default 2 secs i.e. the minimal amount of time allowed between two queries to ES. By setting it to 60 secs, Seb saw a drop to  0.5-1% duplicates. This means that it can take more than a minute between the moment a tuple is acked and when the update is reflected in the search results.

Idea : we could remove the tuples from the beingProcessed hash after some additional time, this way we wouldn't refetch the same URL too soon. The various ES spouts have an overlap in code, we could create an abstract class and share the functionality there.

... and this patch causes the worker to fail sometimes while placing items in the deletion queue/cache:
```
2016-09-26 16:45:37.765 o.a.s.util [ERROR] Async loop died!
java.lang.AssertionError
        at com.google.common.cache.LocalCache$Segment.evictEntries(LocalCache.java:2670) ~[stormjar.jar:?]
        at com.google.common.cache.LocalCache$Segment.put(LocalCache.java:2887) ~[stormjar.jar:?]
        at com.google.common.cache.LocalCache.put(LocalCache.java:4149) ~[stormjar.jar:?]
        at com.google.common.cache.LocalCache$LocalManualCache.put(LocalCache.java:4754) ~[stormjar.jar:?]
        at com.digitalpebble.stormcrawler.elasticsearch.persistence.AbstractElasticSearchSpout$InProcessMap.remove(AbstractElasticSearchSpout.java:84) ~[stormjar.jar:?]
        at com.digitalpebble.stormcrawler.elasticsearch.persistence.AggregationSpout.ack(AggregationSpout.java:398) ~[stormjar.jar:?]
        at org.apache.storm.daemon.executor$ack_spout_msg.invoke(executor.clj:445) ~[storm-core-1.0.1.jar:1.0.1]
...
2016-09-26 16:44:41.827 o.a.s.util [ERROR] Halting process: ("Worker died")
```

--- a/core/src/main/java/com/digitalpebble/stormcrawler/util/MetadataTransfer.java
+++ b/core/src/main/java/com/digitalpebble/stormcrawler/util/MetadataTransfer.java
@@ -22,6 +22,7 @@ import java.util.Set;
 
 import org.apache.commons.lang.StringUtils;
 
+import com.digitalpebble.stormcrawler.Constants;
 import com.digitalpebble.stormcrawler.Metadata;
 
 import clojure.lang.PersistentVector;
@@ -148,6 +149,9 @@ public class MetadataTransfer {
             else {
                 mdToPersistOnly.add(obj.toString());
             }
+
+            // always add the fetch error count
+            mdToPersistOnly.add(Constants.fetchErrorCountParamName);
         }
     }
 

--- a/external/elasticsearch/es-conf.yaml
+++ b/external/elasticsearch/es-conf.yaml
@@ -39,6 +39,9 @@ es.status.routing.fieldname: "hostname"
 es.status.bulkActions: 500
 es.status.flushInterval: "5s"
 
+# used by spouts - time in secs for which the URLs will be considered for fetching after a ack of fail
+es.status.ttl.purgatory: 30
+
 # ElasticSearchSpout
 # ES Spout throttling. Uses configuration of URLPartitionerBolt for the bucket key.
 es.status.max.inflight.urls.per.bucket: -1

--- a/external/elasticsearch/src/main/java/com/digitalpebble/stormcrawler/elasticsearch/persistence/AbstractSpout.java
+++ b/external/elasticsearch/src/main/java/com/digitalpebble/stormcrawler/elasticsearch/persistence/AbstractSpout.java
@@ -17,12 +17,14 @@
 
 package com.digitalpebble.stormcrawler.elasticsearch.persistence;
 
+import java.util.HashMap;
 import java.util.Iterator;
 import java.util.LinkedList;
 import java.util.List;
 import java.util.Map;
 import java.util.Map.Entry;
 import java.util.Queue;
+import java.util.concurrent.TimeUnit;
 
 import org.apache.storm.metric.api.IMetric;
 import org.apache.storm.metric.api.MultiCountMetric;
@@ -42,6 +44,9 @@ import org.slf4j.LoggerFactory;
 import com.digitalpebble.stormcrawler.Metadata;
 import com.digitalpebble.stormcrawler.elasticsearch.ElasticSearchConnection;
 import com.digitalpebble.stormcrawler.util.ConfUtils;
+import com.google.common.base.Optional;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
 
 public abstract class AbstractSpout extends BaseRichSpout {
 
@@ -51,6 +56,12 @@ public abstract class AbstractSpout extends BaseRichSpout {
     protected static final String ESBoltType = "status";
     protected static final String ESStatusIndexNameParamName = "es.status.index.name";
     protected static final String ESStatusDocTypeParamName = "es.status.doc.type";
+
+    /**
+     * Time in seconds for which acked or failed URLs will be considered for
+     * fetching again, default 30 secs.
+     **/
+    protected static final String ESStatusTTLPurgatory = "es.status.ttl.purgatory";
 
     protected String indexName;
     protected String docType;
@@ -71,6 +82,45 @@ public abstract class AbstractSpout extends BaseRichSpout {
     protected String logIdprefix = "";
 
     protected Queue<Values> buffer = new LinkedList<>();
+
+    /**
+     * Map to keep in-process URLs, ev. with additional information for URL /
+     * politeness bucket (hostname / domain etc.). The entries are kept in a
+     * cache for a configurable amount of time to avoid that some items are
+     * fetched a second time if new items are queried shortly after they have
+     * been acked.
+     */
+    protected InProcessMap<String, String> beingProcessed;
+
+    /** Map which holds elements some additional time after the removal. */
+    public class InProcessMap<K, V> extends HashMap<K, V> {
+
+        private Cache<K, Optional<V>> deletionCache;
+
+        public InProcessMap(long maxDuration, TimeUnit timeUnit) {
+            deletionCache = CacheBuilder.newBuilder()
+                    .expireAfterWrite(maxDuration, timeUnit).build();
+        }
+
+        @Override
+        public boolean containsKey(Object key) {
+            boolean incache = super.containsKey(key);
+            if (!incache) {
+                incache = (deletionCache.getIfPresent(key) != null);
+            }
+            return incache;
+        }
+
+        @Override
+        public V remove(Object key) {
+            deletionCache.put((K) key, Optional.absent());
+            return super.remove(key);
+        }
+
+        public long inCache() {
+            return deletionCache.size();
+        }
+    }
 
     @Override
     public void open(Map stormConf, TopologyContext context,
@@ -120,6 +170,11 @@ public abstract class AbstractSpout extends BaseRichSpout {
 
         _collector = collector;
 
+        int ttlPurgatory = ConfUtils
+                .getInt(stormConf, ESStatusTTLPurgatory, 30);
+
+        beingProcessed = new InProcessMap<>(ttlPurgatory, TimeUnit.SECONDS);
+
         eventCounter = context.registerMetric("counters",
                 new MultiCountMetric(), 10);
 
@@ -129,6 +184,21 @@ public abstract class AbstractSpout extends BaseRichSpout {
                 return buffer.size();
             }
         }, 10);
+
+        context.registerMetric("beingProcessed", new IMetric() {
+            @Override
+            public Object getValueAndReset() {
+                return beingProcessed.size();
+            }
+        }, 10);
+
+        context.registerMetric("inPurgatory", new IMetric() {
+            @Override
+            public Object getValueAndReset() {
+                return beingProcessed.inCache();
+            }
+        }, 10);
+
     }
 
     @Override
@@ -160,6 +230,20 @@ public abstract class AbstractSpout extends BaseRichSpout {
             }
         }
         return metadata;
+    }
+
+    @Override
+    public void ack(Object msgId) {
+        LOG.debug("{}  Ack for {}", logIdprefix, msgId);
+        beingProcessed.remove(msgId);
+        eventCounter.scope("acked").incrBy(1);
+    }
+
+    @Override
+    public void fail(Object msgId) {
+        LOG.info("{}  Fail for {}", logIdprefix, msgId);
+        beingProcessed.remove(msgId);
+        eventCounter.scope("failed").incrBy(1);
     }
 
     @Override

--- a/external/elasticsearch/src/main/java/com/digitalpebble/stormcrawler/elasticsearch/persistence/AggregationSpout.java
+++ b/external/elasticsearch/src/main/java/com/digitalpebble/stormcrawler/elasticsearch/persistence/AggregationSpout.java
@@ -19,15 +19,10 @@ package com.digitalpebble.stormcrawler.elasticsearch.persistence;
 
 import java.util.Collections;
 import java.util.Date;
-import java.util.HashSet;
-import java.util.LinkedList;
 import java.util.List;
 import java.util.Map;
-import java.util.Queue;
-import java.util.Set;
 
 import org.apache.commons.lang.StringUtils;
-import org.apache.storm.metric.api.IMetric;
 import org.apache.storm.spout.SpoutOutputCollector;
 import org.apache.storm.task.TopologyContext;
 import org.apache.storm.tuple.Values;
@@ -87,8 +82,6 @@ public class AggregationSpout extends AbstractSpout {
      **/
     private static final String ESStatusMinDelayParamName = "es.status.min.delay.queries";
 
-    protected Set<String> beingProcessed = new HashSet<>();
-
     /** Field name used for field collapsing e.g. metadata.hostname **/
     protected String partitionField;
 
@@ -140,7 +133,7 @@ public class AggregationSpout extends AbstractSpout {
             Values fields = buffer.remove();
 
             String url = fields.get(0).toString();
-            beingProcessed.add(url);
+            beingProcessed.put(url, null);
 
             _collector.emit(fields, url);
             eventCounter.scope("emitted").incrBy(1);
@@ -254,7 +247,7 @@ public class AggregationSpout extends AbstractSpout {
                         hit.getId(), hit.getSourceAsString());
 
                 // is already being processed - skip it!
-                if (beingProcessed.contains(url)) {
+                if (beingProcessed.containsKey(url)) {
                     alreadyprocessed++;
                     continue;
                 }
@@ -282,20 +275,6 @@ public class AggregationSpout extends AbstractSpout {
         eventCounter.scope("already_being_processed").incrBy(alreadyprocessed);
         eventCounter.scope("ES_queries").incrBy(1);
         eventCounter.scope("ES_docs").incrBy(numhits);
-    }
-
-    @Override
-    public void ack(Object msgId) {
-        LOG.debug("{}  Ack for {}", logIdprefix, msgId);
-        beingProcessed.remove(msgId);
-        eventCounter.scope("acked").incrBy(1);
-    }
-
-    @Override
-    public void fail(Object msgId) {
-        LOG.info("{}  Fail for {}", logIdprefix, msgId);
-        beingProcessed.remove(msgId);
-        eventCounter.scope("failed").incrBy(1);
     }
 
 }

--- a/external/elasticsearch/src/main/java/com/digitalpebble/stormcrawler/elasticsearch/persistence/ElasticSearchSpout.java
+++ b/external/elasticsearch/src/main/java/com/digitalpebble/stormcrawler/elasticsearch/persistence/ElasticSearchSpout.java
@@ -22,7 +22,6 @@ import java.util.HashMap;
 import java.util.Map;
 import java.util.concurrent.atomic.AtomicInteger;
 
-import org.apache.storm.metric.api.IMetric;
 import org.apache.storm.spout.SpoutOutputCollector;
 import org.apache.storm.task.TopologyContext;
 import org.apache.storm.tuple.Values;
@@ -78,9 +77,6 @@ public class ElasticSearchSpout extends AbstractSpout {
     /** Keeps a count of the URLs being processed per host/domain/IP **/
     private Map<String, AtomicInteger> inFlightTracker = new HashMap<>();
 
-    // URL / politeness bucket (hostname / domain etc...)
-    private Map<String, String> beingProcessed = new HashMap<>();
-
     // when using multiple instances - each one is in charge of a specific shard
     // useful when sharding based on host or domain to guarantee a good mix of
     // URLs
@@ -108,13 +104,6 @@ public class ElasticSearchSpout extends AbstractSpout {
 
         partitioner = new URLPartitioner();
         partitioner.configure(stormConf);
-
-        context.registerMetric("beingProcessed", new IMetric() {
-            @Override
-            public Object getValueAndReset() {
-                return beingProcessed.size();
-            }
-        }, 10);
     }
 
     @Override
@@ -259,6 +248,7 @@ public class ElasticSearchSpout extends AbstractSpout {
 
     @Override
     public void ack(Object msgId) {
+        LOG.debug("{}  Ack for {}", logIdprefix, msgId);
         String partitionKey = beingProcessed.remove(msgId);
         decrementPartitionKey(partitionKey);
         eventCounter.scope("acked").incrBy(1);
@@ -266,7 +256,7 @@ public class ElasticSearchSpout extends AbstractSpout {
 
     @Override
     public void fail(Object msgId) {
-        LOG.info("Fail for {}", msgId);
+        LOG.info("{}  Fail for {}", logIdprefix, msgId);
         String partitionKey = beingProcessed.remove(msgId);
         decrementPartitionKey(partitionKey);
         eventCounter.scope("failed").incrBy(1);

--- a/external/elasticsearch/src/main/java/com/digitalpebble/stormcrawler/elasticsearch/persistence/SamplerAggregationSpout.java
+++ b/external/elasticsearch/src/main/java/com/digitalpebble/stormcrawler/elasticsearch/persistence/SamplerAggregationSpout.java
@@ -106,7 +106,7 @@ public class SamplerAggregationSpout extends AggregationSpout {
                     hit.getSourceAsString());
 
             // is already being processed - skip it!
-            if (beingProcessed.contains(url)) {
+            if (beingProcessed.containsKey(url)) {
                 alreadyprocessed++;
                 continue;
             }

